// @rpldy/shared-ui/lib/cjs/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "useBatchAddListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useBatchAddListener;
  }
});
Object.defineProperty(exports, "useBatchStartListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useBatchStartListener;
  }
});
Object.defineProperty(exports, "useBatchProgressListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useBatchProgressListener;
  }
});
Object.defineProperty(exports, "useBatchFinishListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useBatchFinishListener;
  }
});
Object.defineProperty(exports, "useBatchCancelledListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useBatchCancelledListener;
  }
});
Object.defineProperty(exports, "useBatchAbortListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useBatchAbortListener;
  }
});
Object.defineProperty(exports, "useItemStartListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useItemStartListener;
  }
});
Object.defineProperty(exports, "useItemFinishListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useItemFinishListener;
  }
});
Object.defineProperty(exports, "useItemProgressListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useItemProgressListener;
  }
});
Object.defineProperty(exports, "useItemCancelListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useItemCancelListener;
  }
});
Object.defineProperty(exports, "useItemErrorListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useItemErrorListener;
  }
});
Object.defineProperty(exports, "useItemAbortListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useItemAbortListener;
  }
});
Object.defineProperty(exports, "useItemFinalizeListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useItemFinalizeListener;
  }
});
Object.defineProperty(exports, "useRequestPreSend", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useRequestPreSend;
  }
});
Object.defineProperty(exports, "useAllAbortListener", {
  enumerable: true,
  get: function get() {
    return _eventListenerHooks.useAllAbortListener;
  }
});
Object.defineProperty(exports, "generateUploaderEventHook", {
  enumerable: true,
  get: function get() {
    return _utils.generateUploaderEventHook;
  }
});
Object.defineProperty(exports, "generateUploaderEventHookWithState", {
  enumerable: true,
  get: function get() {
    return _utils.generateUploaderEventHookWithState;
  }
});
Object.defineProperty(exports, "logWarning", {
  enumerable: true,
  get: function get() {
    return _utils.logWarning;
  }
});
Object.defineProperty(exports, "markAsUploadOptionsComponent", {
  enumerable: true,
  get: function get() {
    return _utils.markAsUploadOptionsComponent;
  }
});
Object.defineProperty(exports, "getIsUploadOptionsComponent", {
  enumerable: true,
  get: function get() {
    return _utils.getIsUploadOptionsComponent;
  }
});
Object.defineProperty(exports, "assertContext", {
  enumerable: true,
  get: function get() {
    return _assertContext["default"];
  }
});
Object.defineProperty(exports, "useUploadOptions", {
  enumerable: true,
  get: function get() {
    return _useUploadOptions["default"];
  }
});
Object.defineProperty(exports, "useAbortItem", {
  enumerable: true,
  get: function get() {
    return _useAbortItem["default"];
  }
});
Object.defineProperty(exports, "useAbortBatch", {
  enumerable: true,
  get: function get() {
    return _useAbortBatch["default"];
  }
});
Object.defineProperty(exports, "useAbortAll", {
  enumerable: true,
  get: function get() {
    return _useAbortAll["default"];
  }
});
Object.defineProperty(exports, "NoDomUploady", {
  enumerable: true,
  get: function get() {
    return _NoDomUploady["default"];
  }
});
Object.defineProperty(exports, "UploadyContext", {
  enumerable: true,
  get: function get() {
    return _UploadyContext["default"];
  }
});
Object.defineProperty(exports, "createContextApi", {
  enumerable: true,
  get: function get() {
    return _UploadyContext.createContextApi;
  }
});
Object.defineProperty(exports, "withRequestPreSendUpdate", {
  enumerable: true,
  get: function get() {
    return _withRequestPreSendUpdate["default"];
  }
});
Object.defineProperty(exports, "useUploadyContext", {
  enumerable: true,
  get: function get() {
    return _useUploadyContext["default"];
  }
});
Object.defineProperty(exports, "useUploady", {
  enumerable: true,
  get: function get() {
    return _useUploadyContext["default"];
  }
});
Object.defineProperty(exports, "getUploadyVersion", {
  enumerable: true,
  get: function get() {
    return _uploadyVersion.getVersion;
  }
});

var _eventListenerHooks = require("./eventListenerHooks");

var _utils = require("./utils");

var _assertContext = _interopRequireDefault(require("./assertContext"));

var _useUploadOptions = _interopRequireDefault(require("./useUploadOptions"));

var _useAbortItem = _interopRequireDefault(require("./useAbortItem"));

var _useAbortBatch = _interopRequireDefault(require("./useAbortBatch"));

var _useAbortAll = _interopRequireDefault(require("./useAbortAll"));

var _NoDomUploady = _interopRequireDefault(require("./NoDomUploady"));

var _UploadyContext = _interopRequireWildcard(require("./UploadyContext"));

var _withRequestPreSendUpdate = _interopRequireDefault(require("./withRequestPreSendUpdate"));

var _useUploadyContext = _interopRequireDefault(require("./useUploadyContext"));

var _uploadyVersion = require("./uploadyVersion");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
