// @rpldy/uploader/lib/cjs/batch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _shared = require("@rpldy/shared");

var _defaults = require("./defaults");

var _utils = require("./utils");

var bCounter = 0;

var processFiles = function (batchId, files, isPending, fileFilter) {
  return Array.prototype.filter.call(files, fileFilter || _defaults.DEFAULT_FILTER).map(function (f) {
    return (0, _shared.createBatchItem)(f, batchId, isPending);
  });
};

var _default = function _default(files, uploaderId, options) {
  bCounter += 1;
  var id = "batch-".concat(bCounter);
  var isFileList = (0, _utils.getIsFileList)(files);
  files = Array.isArray(files) || isFileList ? files : [files];
  var isPending = !options.autoUpload;
  var items = processFiles(id, files, isPending, options.fileFilter);
  return {
    id: id,
    uploaderId: uploaderId,
    items: items,
    state: isPending ? _shared.BATCH_STATES.PENDING : _shared.BATCH_STATES.ADDED,
    completed: 0,
    loaded: 0,
    orgItemCount: items.length
  };
};

exports["default"] = _default;
