// @rpldy/uploader/lib/cjs/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ITEM_FINALIZE_STATES = exports.SENDER_EVENTS = exports.PROGRESS_DELAY = exports.UPLOADER_EVENTS = void 0;

var _shared = require("@rpldy/shared");

var UPLOADER_EVENTS = (0, _shared.devFreeze)({
  BATCH_ADD: "BATCH-ADD",
  BATCH_START: "BATCH-START",
  BATCH_PROGRESS: "BATCH_PROGRESS",
  BATCH_FINISH: "BATCH-FINISH",
  BATCH_ABORT: "BATCH-ABORT",
  BATCH_CANCEL: "BATCH-CANCEL",
  ITEM_START: "FILE-START",
  ITEM_CANCEL: "FILE-CANCEL",
  ITEM_PROGRESS: "FILE-PROGRESS",
  ITEM_FINISH: "FILE-FINISH",
  ITEM_ABORT: "FILE-ABORT",
  ITEM_ERROR: "FILE-ERROR",
  ITEM_FINALIZE: "FILE-FINALIZE",
  REQUEST_PRE_SEND: "REQUEST_PRE_SEND",
  ALL_ABORT: "ALL_ABORT"
});
exports.UPLOADER_EVENTS = UPLOADER_EVENTS;
var PROGRESS_DELAY = 50;
exports.PROGRESS_DELAY = PROGRESS_DELAY;
var SENDER_EVENTS = (0, _shared.devFreeze)({
  ITEM_PROGRESS: "ITEM_PROGRESS",
  BATCH_PROGRESS: "BATCH_PROGRESS"
});
exports.SENDER_EVENTS = SENDER_EVENTS;
var ITEM_FINALIZE_STATES = [_shared.FILE_STATES.FINISHED, _shared.FILE_STATES.ERROR, _shared.FILE_STATES.CANCELLED, _shared.FILE_STATES.ABORTED];
exports.ITEM_FINALIZE_STATES = ITEM_FINALIZE_STATES;
